### PR TITLE
#166880077 User Profile View and Edit

### PR DESCRIPTION
**Pull request by NicholusMuwonge**
#### What does this PR do?

- User should be able to check their profile
- User be able to edit their profile

#### Description of Task to be completed?
- creating a one-page form that is able to retrieve articles and also update articles.
- writing unit tests and global test.

#### How should this be manually tested?
- One needs to `Click Account` on the navbar and get redirected to the profile page route may look like . https://ah-frontend-space-stagin-pr-21.herokuapp.com/profile and when one clicks edit, they get redirected to https://ah-frontend-space-stagin-pr-21.herokuapp.com/editprofile.

#### What are the relevant pivotal tracker stories?
[#166880077](https://www.pivotaltracker.com/story/show/166880077)

Screenshot 2019-07-24 at 21 19 51
Screenshot 2019-07-24 at 21 17 51

### CODE REVIEW

**Review by sebbss — CHANGES_REQUESTED** [review]
@NicholusMuwonge thanks for the implementation. but kindly work on the conflicts with develop

File: src/__test__/actions/profile/ProfileActions.test.js [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    });
+    const store = mockStore({});
+    return store.dispatch(fetchProfile(props))
+      .then(() => {
+        expect(store.getActions()).toEqual(Process);
+      });
+  });
+
+
+  //   it('should be able to return an error on failure to fetch profile', () => {
```
Comment (richien) [inline src/__test__/actions/profile/ProfileActions.test.js]: Hi @NicholusMuwonge, kindly remove this commented out code.

File: src/containers/profile/EditProfile.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        });
+    });
+  }
+
+
+  render() {
+    const {
+      firstName, lastName, dateOfBirth, bio, image, buttonLoading,
+    } = this.state;
+    console.log('STATE.......', image);
```
Comment (richien) [inline src/containers/profile/EditProfile.js]: @NicholusMuwonge , kindly remove these console log functions from your code

Comment (NicholusMuwonge) [reply src/containers/profile/EditProfile.js]: I have reacted to this please

File: src/__test__/actions/profile/ProfileActions.test.js (hunk 1 above)
Comment (NicholusMuwonge) [reply src/__test__/actions/profile/ProfileActions.test.js]: I have reacted to this please

**Comment by NicholusMuwonge** [comment]
> @NicholusMuwonge thanks for the implementation. but kindly work on the conflicts with develop

Sir, I have reacted to your feedback, kindly check back please

**Review by frankopkusianwar — COMMENTED** [review]
@NicholusMuwonge this is good work, however I would recommend you add a profile image avatar as a placeholder before someone can update their profile picture

**Comment by NicholusMuwonge** [comment]
> @NicholusMuwonge this is good work, however I would recommend you add a profile image avatar as a placeholder before someone can update their profile picture

I have reacted to that feedback. Thanks

**Comment by NicholusMuwonge** [comment]
> @NicholusMuwonge this is good work, however I would recommend you add a profile image avatar as a placeholder before someone can update their profile picture

I have reacted to this feedback please

**Review by richien — APPROVED** [review]

**Review by frankopkusianwar — APPROVED** [review]
LGTM

**Review by sebbss — APPROVED** [review]

File: src/__test__/actions/firstAction.test.js [hunk 3]
```diff
@@ -1,5 +1,15 @@
 import FirstAction from '../../actions/FirstAction';
 import { trialType } from '../../actions/types';
+import Index from '../../index';
+
+jest.mock('react-dom', () => ({ render: jest.fn() }));
+
+it('renders without crashing', () => {
```
Comment (MuhanguziDavid) [inline src/__test__/actions/firstAction.test.js]: put this test in its own file

File: src/__test__/actions/profile/ProfileActions.test.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+      },
+    };
+
+    const props = {
+      history: {
+        push: jest.fn(),
+      },
+    };
+
+
```
Comment (MuhanguziDavid) [inline src/__test__/actions/profile/ProfileActions.test.js:152]: please remove the extra line

File: src/__test__/components/profile/ProfileEdit.test.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  it('it should map state to props', () => {
+    const initialState = {
+      profile: '',
+    };
+    expect(mapStateToProps(initialState)).toEqual({
+      profile: '',
+    });
+  });
+
+  it('should handle change', () => {
```
Comment (MuhanguziDavid) [inline src/__test__/components/profile/ProfileEdit.test.js]: use more descriptive test names

File: src/__test__/components/profile/ProfileEdit.test.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        value: 'first',
+      },
+    };
+    const wrapperInst = wrapper.instance();
+    wrapperInst.handleChange(event);
+    wrapperInst.handleSubmit(event);
+    expect(wrapperInst.state.first_name).toBe('first');
+  });
+
+  it('should setState ', () => {
```
Comment (MuhanguziDavid) [inline src/__test__/components/profile/ProfileEdit.test.js]: use more descriptive test names

File: src/__test__/reducers/profile/ProfileReducer.test.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+let error;
+let retrievedProfile;
+
+describe('profileReducer', () => {
+  it(('it tests the reducer initial state'), () => {
+    const newState = profileReducer(initialstate, {});
+    expect(newState).toEqual(initialstate);
+  });
+
+  it('test that loading occurs', () => {
```
Comment (MuhanguziDavid) [inline src/__test__/reducers/profile/ProfileReducer.test.js]: It is already clear that this is a test. You do not have to use the word `test`
Please use something like `it('should load...'`.

Comment (MuhanguziDavid) [reply src/__test__/reducers/profile/ProfileReducer.test.js]: Kindly please, sir

File: src/actions/profileActions.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    .then((response) => {
+      dispatch(loadProfile(response.data));
+    })
+    .catch((error) => {
+      dispatch(failed(error.response.data));
+      props.history.push('/login');
+    });
+};
+
+export const editUserProfile = (profileInput, props) => async dispatch => axios.put(`https://ah-haven-space.herokuapp.com/api/profiles/${(sessionStorage.getItem('username'))}/edit`, profileInput, {
```
Comment (MuhanguziDavid) [inline src/actions/profileActions.js]: this line is too long. Please refactor.

File: src/assets/css/Profile.scss [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  text-align: center;
+}
+
+.profile {
+  padding-bottom: 58px;
+}
+
+.edit-button {
+    background-color: #00bfff;
+}
```
Comment (MuhanguziDavid) [inline src/assets/css/Profile.scss]: add a line at the end of this file!

File: src/components/profile/ProfileEdit.js [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    <Fragment>
+      <div className="buttons">
+
+        <button type="submit" onClick={onSubmit} disabled={buttonLoading} className="btn btn-primary btn-lg">
+          {buttonLoading && (<i className="spinner-grow spinner-grow-sm" />)}
+          {buttonLoading && <span>Just A sec Please!...</span>}
+          {!buttonLoading && <span>Update Profile</span>}
+        </button>
+        <Link className="btn btn-danger btn-lg cancel" to="/profile">
+                Cancel
```
Comment (MuhanguziDavid) [inline src/components/profile/ProfileEdit.js:18]: please fix the indentation of this line

File: src/components/profile/ProfileFetch.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+            >
+              {(profile.image !== 'url') ? (
+                <img
+                  src={profile.image}
+                  alt=""
+                  width="370em"
+                  height="310em"
+                />
+              )
+                : (
```
Comment (MuhanguziDavid) [inline src/components/profile/ProfileFetch.js]: the indentation of this block and the previous block should be the same. Please refactor

File: src/containers/profile/EditProfile.js [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    });
+    fileload.on('state_changed', (snapshot) => {
+      const status = Math.round(
+        (snapshot.bytesTransferred / snapshot.totalBytes) * 100,
+      );
+      this.setState({ status });
+    });
+  }
+
+
```
Comment (MuhanguziDavid) [inline src/containers/profile/EditProfile.js]: please remove the extra line of code

File: src/__test__/actions/firstAction.test.js (hunk 3 above)
Comment (NicholusMuwonge) [reply src/__test__/actions/firstAction.test.js]: This is reacted to please.

File: src/__test__/actions/profile/ProfileActions.test.js (hunk 4 above)
Comment (NicholusMuwonge) [reply src/__test__/actions/profile/ProfileActions.test.js:152]: reacted to please

File: src/__test__/components/profile/ProfileEdit.test.js (hunk 5 above)
Comment (NicholusMuwonge) [reply src/__test__/components/profile/ProfileEdit.test.js]: I have kindly reacted to this

File: src/__test__/components/profile/ProfileEdit.test.js (hunk 6 above)
Comment (NicholusMuwonge) [reply src/__test__/components/profile/ProfileEdit.test.js]: This is also reacted to please

File: src/__test__/reducers/profile/ProfileReducer.test.js (hunk 7 above)
Comment (NicholusMuwonge) [reply src/__test__/reducers/profile/ProfileReducer.test.js]: Reacted to this please

File: src/actions/profileActions.js (hunk 8 above)
Comment (NicholusMuwonge) [reply src/actions/profileActions.js]: this has been fixed

**Review by MuhanguziDavid — APPROVED** [review]
LGTM

**Pull request merged**